k8s.gcr.io/lib.sh

```diff
@@ -0,0 +1,279 @@
+        gcloud --project "${project}" \
+            container images delete --quiet "${dest}:latest"
+    fi
+
+    gsutil iam ch allUsers:objectViewer $(gcs_bucket_for "${project}" "${region}")
+}
+
+# Grant full privileges to GCR admins
+# $1: The GCP project
+function empower_gcr_admins() {
+    if [ $# -lt 1 -o $# -gt 2 -o -z "$1" ]; then
+        echo "empower_gcr_admins(project, [region]) requires 1 or 2 arguments" > /dev/stderr
+        return 1
+    fi
+    project="$1"
+    region="${2:-}"
+
+    # Grant project viewer so the UI will work.
+    gcloud \
+        projects add-iam-policy-binding "${project}" \
+        --member "group:${GCR_ADMINS}" \
+        --role roles/viewer
+
+    # Grant admins access to do admin stuff.
+    gsutil iam ch \
+        "group:${GCR_ADMINS}:objectAdmin" \
+        $(gcs_bucket_for "${project}" "${region}")
+    gsutil iam ch \
+        "group:${GCR_ADMINS}:legacyBucketOwner" \
+        $(gcs_bucket_for "${project}" "${region}")
+}
+
+# Grant write privileges to a group
+# $1: The GCP project
+# $2: The googlegroups group
+function empower_group() {
+    if [ $# -lt 2 -o $# -gt 3 -o -z "$1" -o -z "$2" ]; then
+        echo "empower_group(project, group_name, [region]) requires 2 or 3 arguments" > /dev/stderr
+        return 1
+    fi
+    project="$1"
+    group="$2"
+    region="${3:-}"
+
+    gsutil iam ch \
+        "group:${group}:objectAdmin" \
+        $(gcs_bucket_for "${project}" "${region}")
+    gsutil iam ch \
+        "group:${group}:legacyBucketReader" \
+        $(gcs_bucket_for "${project}" "${region}")
+}
+
+# Grant full privileges to the GCR promoter bot
+# $1: The GCP project
+function empower_promoter() {
+    if [ $# -lt 1 -o $# -gt 2 -o -z "$1" ]; then
+        echo "empower_promoter(project, [region]) requires 1 or 2 arguments" > /dev/stderr
+        return 1
+    fi
+    project="$1"
+    region="${2:-}"
+
+    acct=$(svc_acct_for "${project}" "${PROMOTER_SVCACCT}")
+
+    if ! gcloud --project "${project}" iam service-accounts describe "${acct}" >/dev/null 2>&1; then
+        gcloud --project "${project}" \
+            iam service-accounts create \
+            "${name}" \
+            --display-name="k8s-infra container image promoter"
+    fi
+
+    # Grant admins access to do admin stuff.
+    gsutil iam ch \
+        "serviceAccount:${acct}:objectAdmin" \
+        $(gcs_bucket_for "${project}" "${region}")
+    gsutil iam ch \
+        "serviceAccount:${acct}:legacyBucketOwner" \
+        $(gcs_bucket_for "${project}" "${region}")
+}
+
+# Configure bucket lifecycle for a prod repo
+# $1: The GCP project
+function ensure_prod_lifecycle() {
+    if [ $# -lt 1 -o $# -gt 2 -o -z "$1" ]; then
+        echo "ensure_prod_lifecycle(project, [region]) requires 1 or 2 arguments" > /dev/stderr
+        return 1
+    fi
+    project="$1"
+    region="${2:-}"
+
+    # Set lifecycle policies.
+    # TODO: do we want this?  It will inhibit promoter's GC, but will protect
```

Comment:
I don't think the retention policy is the right safeguard here. Right off the bat it would prevent us from performing reverts against mistakenly-promoted images (I've already seen this happen on more than one occasion for k8s.gcr.io).

For pruning/deleting old images, it would be better to have a script that periodically runs and looks for old images in the promoter manifest, and make it file a PR to delete that image from the manifest. It would make everything explicit.

On a related note, I need to add a feature of the promoter that "snapshots" a GCR into YAML (create a promoter manifest that captures the existing state).

Reply:
Yeah, I am really on the fence.  

When we promote a bad image, what do we want to happen?  Do we really need to delete it, or just change the tag back?

I worry that this is a vector for mistakes that retention policy (especially if locked!!) could prevent.   E.g. maybe we don't need a graveyard, just a GC and very long minimum retention.  The behavior I really want (delayed delete) isn't part of GCS.  I think I'd rather pay the 3-year or 5-year cost of hosting something that was accidentally promoted if it means I know that real images can't go away for literally years.

What we really want is immutable labels.  I don't think we have a way to specify that, either.


(and yes, we will need a dump tool to import the thousands of existing images)

Reply:
I guess setting a GCS retention policy does not prevent us from moving around/renaming Docker tags? If so, then I think having a retention policy is fine.

The only thing I don't like is how we are relinquishing control of image deletion to the underlying GCS layer beneath GCR. This is something the promotion/demotion process can handle, so why put this piece in?

Also, this prevents us from performing real e2e tests around promotion using these exact GCRs in the future (we won't be able to delete test data). Maybe this seems far-fetched, but in my mind I think it is useful to have real tests that run on real infrastructure --- just in case the real one isn't behaving as expected for whatever reason.

I vote to add in a retention policy later after thinking about it some more. I feel like it should be a more thought-out decision.

Reply:
Why: belts and suspenders.  k8s is too important to risk.  Consider what happens if a bug nukes an image that people use.  The impact is too big to ignore.

We could e2e to alternate places, that seems ok.

I am fine to not do this for now, but I *do* want to think about it. 

I will let it sit for the weekend.

Reply:
For people coming into the discussion from the GH side, the ML thread is: https://groups.google.com/forum/#!topic/kubernetes-wg-k8s-infra/Dz4urYxWKi4